proxy/common/backports.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+"""
+    proxy.py
+    ~~~~~~~~
+    ⚡⚡⚡ Fast, Lightweight, Pluggable, TLS interception capable proxy server focused on
+    Network monitoring, controls & Application development, testing, debugging.
+
+    :copyright: (c) 2013-present by Abhinav Singh and contributors.
+    :license: BSD, see LICENSE for more details.
+"""
+import time
+
+from typing import Any
+
+
+class cached_property:
```

Comment:
I don't think you should vendor this. It's more sustainable to just use a common pattern of falling back to use a backport: https://docs.python.org/3/library/functools.html#functools.cached_property + https://pypi.org/p/backports.cached-property.

Reply:
I thought about it.  Instead of a `pip install` I'll prefer a baked in implementation.  I picked this up from some StackOverflow thread.  But point being, we can further customize and build on top of this base implementation for `proxy.py` needs.  I don't think we'll be limited to a simple `@cached_property`, we'll likely need caching strategies and such in future.  Then, we'll able to modify this to our taste.

Reply:
Vendoring things may be undesired from a legal PoV. In that case you'd have to comply with the license requirements which often demand that the original code is distinguishable from that you add. This is why vendored libs usually live in their own directory, in separate files/subfolders, and you'll have to maintain their licenses+legal notices in prominent places. Plus, adding your modifications on top would need to be clean, probably separate, and/or well-documented.
Also, `pip install` already gets all the deps and is a de-facto standard.
Vendoring implies added maintenance burden: you'll have to manually figure out if the upstream project has any security vulnerabilities, apply fixes and re-release your project quickly (because otherwise, you steal the ability to get security fixes from your users). This is non-automatable, especially not with pieces of code that is not distributed in a standard way. You could solve this by integrating the tool that pip itself uses (and only because it technically cannot rely on the deps for self-bootstapping): https://pypi.org/p/vendoring.
This should be meant as the last resort and all the projects using it eventually move away from this technique whenever possible.
At least two projects I'm involved with that used to rely on vendored deps have moved away from them when it became possible (CherryPy and Ansible).
With vendoring, you'll also run into problems in distro packaging ecosystems, in particular, because they patch projects and de-vendor their deps. For example, if Fedora has a dependency that is already packaged separately, they'd mark it as your dependency, remove your vendored code and apply patches so that your code imports that.
What I suggested above is a common and well-understood way of having shims for features in stdlib with backports available.